src/client/shadows/dynamicshadows.cpp

```diff
@@ -0,0 +1,170 @@
+	radius = boundVec.getLength() * 2.0f;
+	// boundVec.getLength();
+	float vvolume = radius * 2.0f;
+
+	float texelsPerUnit = getMapResolution() / vvolume;
+	m4f mTexelScaling;
+	mTexelScaling.setScale(texelsPerUnit);
+
+	m4f mLookAt, mLookAtInv;
+
+	mLookAt.buildCameraLookAtMatrixLH(v3f(0.0f, 0.0f, 0.0f), -direction, v3f(0.0f, 1.0f, 0.0f));
+
+	mLookAt *= mTexelScaling;
+	mLookAtInv = mLookAt;
+	mLookAtInv.makeInverse();
+
+	v3f frustumCenter = newCenter;
+	mLookAt.transformVect(frustumCenter);
+	frustumCenter.X = floorf(frustumCenter.X); // clamp to texel increment
+	frustumCenter.Y = floorf(frustumCenter.Y); // clamp to texel increment
+	frustumCenter.Z = floorf(frustumCenter.Z);
+	mLookAtInv.transformVect(frustumCenter);
+	// probar radius multipliacdor en funcion del I, a menor I mas multiplicador
+	v3f eye_displacement = direction * vvolume;
+
+	// we must compute the viewmat with the position - the camera offset
+	// but the shadow_frustum position must be the actual world position
+	v3f eye = frustumCenter - eye_displacement;
+	shadow_frustum.position = world_center - eye_displacement;
+	shadow_frustum.length = vvolume;
+	shadow_frustum.ViewMat.buildCameraLookAtMatrixLH(eye, frustumCenter, v3f(0.0f, 1.0f, 0.0f));
+	shadow_frustum.ProjOrthMat.buildProjectionMatrixOrthoLH(shadow_frustum.length,
+			shadow_frustum.length, -shadow_frustum.length,
+			shadow_frustum.length,false);
+}
+
+DirectionalLight::DirectionalLight(const u32 shadowMapResolution,
+		const v3f &position, video::SColorf lightColor,
+		f32 farValue) :
+		diffuseColor(lightColor),
+		farPlane(farValue), mapRes(shadowMapResolution), pos(position)
+{}
+
+void DirectionalLight::update_frustum(const Camera *cam, Client *client)
+{
+	should_update_map_shadow = true;
+	float zNear = cam->getCameraNode()->getNearValue();
+	float zFar = getMaxFarValue();
+
+	///////////////////////////////////
+	// update splits near and fars
+	shadow_frustum.zNear = zNear;
+	shadow_frustum.zFar = zFar;
+
+	// update shadow frustum
+	createSplitMatrices(cam);
+	// get the draw list for shadows
+	client->getEnv().getClientMap().updateDrawListShadow(
+			getPosition(), getDirection(), shadow_frustum.length);
+	should_update_map_shadow = true;
+}
+
+void DirectionalLight::setDirection(v3f dir)
+{
+	direction = -dir;
+	direction.normalize();
+}
+
+v3f DirectionalLight::getDirection() const
+{
+	return direction;
+}
+
+v3f DirectionalLight::getPosition() const
+{
+	return shadow_frustum.position;
+}
+
+const m4f &DirectionalLight::getViewMatrix() const
+{
+	return shadow_frustum.ViewMat;
+}
+
+const m4f &DirectionalLight::getProjectionMatrix() const
+{
+	return shadow_frustum.ProjOrthMat;
+}
+
+m4f DirectionalLight::getViewProjMatrix()
+{
+	return shadow_frustum.ProjOrthMat * shadow_frustum.ViewMat;
+}
+
+f32 DirectionalLight::getMaxFarValue() const
+{
+	return farPlane;
+}
+
+const video::SColorf &DirectionalLight::getLightColor() const
+{
+	return diffuseColor;
```

Comment:
this can be inlined in class def

Reply:
I'm not a big fan of code in headers, but it's ok :)